c7n/actions.py

```diff
@@ -522,3 +571,102 @@ def process(self, resources, event):
             {'key': self.data['tag'], 'value': user},
             self.manager).process(untagged)
         return untagged
+
+
+class PutMetric(BaseAction):
+    """Action to put metrics based on an expression into CloudWatch metrics
+    
+    :example:
+    
+        .. code-block: yaml
+        
+            policies:
+              - name: track-attached-ebs
+                resource: ec2
+                comment: |
+                  Put the count of the number of EBS attached disks to an instance
+                filters:
+                  - Name: tracked-ec2-instance
+                actions:
+                  - type: put-metric
+                    key: Reservations[].Instances[].BlockDeviceMappings[].DeviceName
+                    namespace: Usage Metrics
+                    metric_name: Attached Disks
+                    op: count
+                    units: Files
+
+    op and units are optional and will default to simple Counts.
+    """
+    permissions = {'cloudwatch:PutMetricData',} #  typically lowercase servicename:TitleCaseActionName
+    schema = {
+        'type': 'object',
+        'required': ['type', 'key', 'namespace', 'metric_name'],
+        'properties': {
+            'type': {'enum': ['put-metric', ]},
+            'key': {'type': 'string'},  # jmes path
+            'namespace': {'type': 'string'},
+            'metric_name': {'type': 'string'},
+            'op': {'enum': METRIC_OPS.keys()},
+            'units': {'enum': METRIC_UNITS}
+        }
+    }
+
+    def process(self, resources):
+        ns = self.data['namespace']
+        metric_name = self.data['metric_name']
+        key_expression = self.data.get('key', 'Resources[]')
+        operation = self.data.get('op', 'count')
+        units = self.data.get('units', 'Count')
+
+        now = datetime.utcnow()
+
+        # reduce the resources by the key expression, and apply the operation to derive the value
+        values = []
+        self.log.debug( "searching for %s in %s", key_expression, resources )
+        try:
+            values = jmespath.search("Resources[]." + key_expression, {'Resources': resources})
+            # I had to wrap resourses in a dict like this in order to not have jmespath expressions start with []
+            # in the yaml files.  It fails to parse otherwise.
+        except TypeError, oops:
+            self.log.error( oops.message )
+
+        value = 0
+        try:
+            f = METRIC_OPS[operation]
+            value = f(values)
+        except KeyError, bad_op:
+            self.log.error( "Bad op for put-metric action: %s", operation )
+
+        # for demo purposes
+        #from math import sin, pi
+        #value = sin((now.minute * 6 * 4 * pi) / 180) * ((now.hour + 1) * 4.0)
+
+
+        metrics_data = [
+            {
+                'MetricName': metric_name,
+                # TODO: support custom dimensions
```

Comment:
OK.   I've done this so that they can be passed like:

```yaml
Dimensions:
- {a: b, c: d}
```
or 
```yaml
Dimensions:
- {a: b}
- {c: d}
```